initial implementation of a github actions reporter

devoncarew opened the pull request:
- initial implementation of a github actions reporter
- address https://github.com/dart-lang/test/issues/1415

Creating as a draft PR for now - I want to play with the output some.

jakemac53 commented on pkgs/test_core/lib/src/runner/reporter/github.dart:
```diff
@@ -0,0 +1,196 @@
+    _sink.writeln();
+
+    final hadFailures = _engine.failed.isNotEmpty;
+    String message =
```
Probably doesn't matter in practice but we should use a StringBuffer here

devoncarew replied:
Updated to use a StringBuffer

jakemac53 commented on pkgs/test_core/lib/src/runner/reporter/github.dart:
```diff
@@ -0,0 +1,196 @@
+    if (isLoadSuite) {
+      // Don't emit any info for 'loadSuite' tests, unless they contain errors.
+      if (errors.isNotEmpty || messages.isNotEmpty) {
+        _sink.writeln('${test.suite.path}:');
```
Should we also be visualizing this as a group? If not why not?

devoncarew replied:
I believe this 'LiveTest' is just plumbing used by package:test during test discovery - it's not an actual user test, and reporting it as such would be odd for the user.

We also special case these LoadSuite test in the other reporters I believe (but don't have the code in front of me right now).

jakemac53 replied:
So what this is specifically is a synthetic "test" used to report the success or failure of compiling, loading, and connecting to the test. Basically, it fails if there is a compilation error in the test or some failure connecting to it.

We do want to surface it for sure, and it looks a bit weird imo to not have it show up the same way (I think it makes sense to have the actual errors hidden behind a collapsable group still for instance).

devoncarew replied:
So a `Loading test/message_extraction/failed_extraction_test.dart` group, which would generally be empty, but when it had content would be important to display (esp. as likely no other tests would run).

I wonder if we shouldn't hide it normally - as mostly it'll be empty and be noise in the rest of the output.

jakemac53 replied:
If we only show it at all when it fails, then I would call it `Failed to load test/message_extraction/failed_extraction_test.dart`. But ya that SGTM

devoncarew replied:
I've updated this to show the load suite info if it has any content - messages or errors; it shows as a group, w/ a different icon from a test.

devoncarew commented on pkgs/test_core/lib/src/runner/reporter/github.dart:
```diff
@@ -113,6 +114,9 @@ class GithubReporter implements Reporter {
       }
     }
 
+    // TODO: how to recognize (setUpAll) and (tearDownAll)?
```
Do these show up in the other test reporters? I really don't remember them displaying in test results.

I think it's odd - if people have 3 tests defined - to display results for 5 tests. Are the results for these included in the `engine.passed` set of tests? Again, I think it would be odd - if they have 3 tests defined in a project - for the test run to conclude with `All 5 tests passed`.

Perhaps a setup and teardown phase can be visualized a bit differently from a regular test?

▶ test/message_extraction/failed_extraction_test.dart: foo.setUpAll()
▶ test/message_extraction/failed_extraction_test.dart: foo.tearDownAll()

jakemac53 replied:
> Do these show up in the other test reporters? I really don't remember them displaying in test results.

They do show up today yes (you would usually only see them if they fail, like all other tests in the compact reporter). On github (expanded reporter) you definitely see them though (something like `00:00 +0: (setUpAll)`).

> I think it's odd - if people have 3 tests defined - to display results for 5 tests. 

Yes, it is a bit odd and the current reporters _don't_ count these in the total number of "successes". It gets weirder if they fail, they do count as a failure, and then no other tests run at all so it reports one test failing regardless of how many tests are in the suite (the others disappear basically).

> Perhaps a setup and teardown phase can be visualized a bit differently from a regular test?

I am fine with them being visualized differently, to some extent this already happens (the name is `(setUpAll)` in the example runs here). Not sure if it would also show the group name, but that would be good.

devoncarew replied:
From some experimentation, setUp() and tearDown() don't cause new tests to be created; setUpAll() and tearDownAll() do.

They're comparatively rare, so I think what we do here isn't critical. I'll likely update this to display them using groups, and make sure the group doesn't look like a normal, user-authored test.

jakemac53 replied:
> From some experimentation, setUp() and tearDown() don't cause new tests to be created; setUpAll() and tearDownAll() do.

Ya, because these run _per test_ they are considered just a part of the test. The thing with `setUpAll` is it runs once, before all the tests, and so you can't link it to any one of them. I guess you could link it to _all_ of them though? But I think that just ends up being noisy, would duplicating the same error in each test.

devoncarew replied:
Gotcha - thanks for the background, makes sense.

kevmoo commented on pkgs/test_core/lib/src/runner/reporter/github.dart:
```diff
@@ -0,0 +1,198 @@
+
+class _GithubHelper {
+  // Char sets avilable at https://www.compart.com/en/unicode/.
+  static const String passed = '✅';
```
Soon this could be an enum! 🤤

devoncarew commented:
OK, I think this is ready for review now. Here's the current output:
- everything (tests, loading diagnostics (loadSuite), setUpAll and tearDownAll methods get wrapped in groups
- if there's no content (errors or messages) in the synthetic (non-test) groups, they're not emitted
- test successes have ✅ icons; failures have ❌ icons
- synthetic groups that pass use icons that are different from tests so they're not confused with user authored tests; ⏺
- skipped tests have ❎ icons; I played with many different available icons here - this seemed like the best; many others were false starts - (on github were rendered w/ a red color - indicating a failure, in a different width than the other icons, making the left edge of the output ragged, ...)
- we print a success message at the end with 🎉; if a failure, we use the github `::error::` output

Some pics for the output below; others can be examined from the 'Checks' tab.

The tests for this were adopted from the ones for the other reporters. Some common ones were removed - ones that tested other functionality but that used the reporter to do that (the expanded and compact reporters still have these tests). And tests added for the github specific UI.

One thing likely to address before landing is that because of the way we collect output until test is complete, we can re-sequence the order of stdout and error messages.

<img width="453" alt="Screen Shot 2022-03-20 at 10 23 49 AM" src="https://user-images.githubusercontent.com/1269969/159581960-cc91d494-180b-4b16-9e8b-2c814e5fd818.png">

<img width="427" alt="Screen Shot 2022-03-20 at 10 23 13 AM" src="https://user-images.githubusercontent.com/1269969/159581996-d862d463-55bb-44e6-982e-607c4579c8a8.png">

<img width="425" alt="Screen Shot 2022-03-20 at 10 22 48 AM" src="https://user-images.githubusercontent.com/1269969/159582046-5c907096-4a64-45a0-aa25-bd341eaf508d.png">

<img width="513" alt="Screen Shot 2022-03-17 at 9 30 11 AM" src="https://user-images.githubusercontent.com/1269969/159582154-53e15ba9-4759-414f-8a42-0e395f8afa5c.png">

devoncarew commented:
This reporter also doesn't have the:

```
Consider enabling the flag chain-stack-traces to receive more detailed exceptions.
For example, 'dart test --chain-stack-traces'.
```

message at the end of every run (I'm not quite sure what that is, or if there's not some other way to let users know).

natebosch commented on pkgs/test_core/lib/src/runner/reporter/github.dart:
```diff
@@ -0,0 +1,198 @@
+
+class _GithubHelper {
+  // Char sets avilable at https://www.compart.com/en/unicode/.
+  static const String passed = '✅';
```
[nit] I have a slight preference for `'\u2705'`, but I'm also curious what other folks think.

@grouma @jakemac53 @lrhn ?

kevmoo replied:
`'\u2705'` is NOT super readable 🤷  – why the preference?

natebosch replied:
Historically they've not worked well in some editors and they aren't easy to type.
Mostly we just haven't used them in source strings that I recall, and I've only seen the `\u` version used.

devoncarew replied:
I changed these to use the unicode escaping form.

grouma replied:
Sorry for the late response. I think I prefer `\u2705` as a constant variable with an appropriate name, e.g. `_checkMark`. That way it's readable and will work with all tools / editors.

natebosch replied:
One point against the `'\u'` form is that the Java and JavaScript style guides at google both seem to encourage it when it's "easier to read", which I think does apply here.
https://google.github.io/styleguide/javaguide.html#s2.3.3-non-ascii-characters
https://google.github.io/styleguide/jsguide.html#non-ascii-characters

C++ mentions that no user facing strings should be in source, which isn't really a practice for us, but they do still specifically allow non-ascii in source.
https://google.github.io/styleguide/cppguide.html#Non-ASCII_Characters

I'm very willing to accept that my preference is old-fashioned and choose a more modern style for our source code if that's the consensus - I mostly just wanted to double check that we weren't introducing a new pattern that would become standard without thinking it through.

@munificent - do you think this is something we should cover in Effective Dart?

jakemac53 replied:
Fwiw for me having the actual characters is better, I have no idea what `\u2705` is and would have to look it up. I don't really see any downside to using the actual character and you have the advantage of seeing it visually.

munificent replied:
> @munificent - do you think this is something we should cover in Effective Dart?

Seems pretty nit-picky to me, but if you think it's worth it, I'm persuadable.

devoncarew replied:
I think - for readability of this particular piece of code - I'm going to switch back to having the actual char in the string. It's not causing issues w/ tooling, and makes it much easier to read this bit od code.

munificent replied:
FWIW, I'm generally comfortable with having non-ASCII characters inside Dart comments and string literals. It's entirely well-specified by the language and there's nothing intrinsically better about Unicode codepoints whose value happens to be less than 128. (I, of course, avoid homoglyphs and other nasty ones.)

natebosch commented on pkgs/test_core/lib/src/runner/reporter/github.dart:
```diff
@@ -0,0 +1,198 @@
+      }
+    }
+    _sink.writeln(_helper.startGroup('$prefix $name$statusSuffix'));
+    // TODO: Note that printing messages and errors in this manner could display
+    // them in a different order than they were generated.
```
Do you mean specifically that messages and errors won't be interleaved, not that messages could be out of order relative to other messages?

I think that's probably fine.

I'm curious though if we should try to do anything with tests that print after they are complete - do I understand correctly that they'd get lost?

devoncarew replied:
The messages will be in the right order, and the errors will be in the right order, but this might interleave the two types differently than how they'd been created.

devoncarew replied:
> I'm curious though if we should try to do anything with tests that print after they are complete - do I understand correctly that they'd get lost?

No - we specifically handle that case (before getting the associated test to pass we had been dropping them). We now track when tests are completed; if we get output for a test which has already completed, we just emit the output directly to the sink (w/o any batching). The output will appear between test groups - it may not be clear what test it's from, but at least it won't be lost.

kevmoo reviewed (changes requested):
Should update the changelog describing the change

Should also add something to the readme explaining how GitHub CI detection is handled – so folks aren't confused.

Looks like it's just auto-magically changed default (👍 ), so it'd be good to document that the user can override the behavior by being explicit w/ the `--reporter` option

devoncarew commented:
> Should update the changelog describing the change
> 
> Should also add something to the readme explaining how GitHub CI detection is handled – so folks aren't confused.
> 
> Looks like it's just auto-magically changed default (👍 ), so it'd be good to document that the user can override the behavior by being explicit w/ the `--reporter` option

Makes sense - I updated the changelog and added a section about test reporters to the readme.

I also rev'd the version of package:test to a pre-release version.

devoncarew commented:
In terms of the package:test version, I rev'd this from `1.20.2` to `1.21.0-dev`.

kevmoo commented on pkgs/test/README.md:
```diff
@@ -233,6 +236,19 @@ dart test --test-randomize-ordering-seed=random
+You can adjust the output format of the test results using the `--reporter=<option>`
+command line flag. The default format is the `compact` output format - a single
+line, continuously updated as tests are run. When running on the GitHub Actions CI
+however, the default changes to the `github` output format - a reporter customized
```
Pondering if we should be explicit abouth the ENV var we check. So it's clear it's not magic.

devoncarew commented:
> In terms of the package:test version, I rev'd this from `1.20.2` to `1.21.0-dev`.

cc @natebosch, @jakemac53 re: the version we should use here

jakemac53 commented:
Minor version bump SGTM.

kevmoo commented on pkgs/test_core/lib/src/runner/reporter/github.dart:
```diff
@@ -0,0 +1,196 @@
+  }
+}
+
+class _GithubHelper {
```
Why not just make this a `abstract` class w/ everything static?

Planning on having state at some point?

jakemac53 replied:
Or just make these top level private members

devoncarew replied:
I'd like to keep these grouped with the other github markup code (`startGroup()`, `error()`, ...) as they're related. No real options other than that; `static` members works for me.

jakemac53 commented on pkgs/test/test/runner/github_reporter_test.dart:
```diff
@@ -0,0 +1,346 @@
+            tearDownAll(() { print('two'); });
+            test('test 1', () {});
+          });''', '''
+          ::group::⏺ one (setUpAll)
```
Why not use ✅ for these as well? This looks to me like a warning indicator.

I would consider using this icon for skipped tests actually?

devoncarew replied:
There's some (incomplete) discussion of the icons here: https://github.com/dart-lang/test/pull/1678#issuecomment-1075677817. I did play around w/ a lot of different icons, including just editing the dom of the 'checks' page to see how things would look.

For this icon, I don't want to use the same icon as for a regular test; I think it's valuable for 6 users test ==> 6 test icons in the output. I'd like the additional test plumbing to have distinct icons from the passed tests. Additionally, the icon should be the same width as the other icons - so the LHS isn't ragged - and shouldn't use a color which users already attach some meaning to (red, yellow, ...).

Other similar unicode chars can be found here: https://unicode-table.com/en/blocks/dingbats/.

jakemac53 replied:
What about ☑? That way it is different but still indicates success (it is a test that can fail). And it isn't a color that indicates a warning. My main complaint about the current choice is the orange color indicates a warning to me.

devoncarew replied:
I've been checking the rendering in-line in the test results, just so we're getting the font that github uses there. What I see for the current 'non-test' group char (⏺) is:

<img width="116" alt="Screen Shot 2022-03-29 at 9 45 09 AM" src="https://user-images.githubusercontent.com/1269969/160663421-91050ddb-aab3-4de5-9db5-dd444d6b4997.png">

Which - for me - is a gray color, not an orange color. I do agree that orange (or red, or yellow) would indicate a failure or problem of some kind to the user - something we should avoid unless there's an actual issue.

Here's what the ☑ char looks like in-line in test results (note the different size and style):

<img width="201" alt="Screen Shot 2022-03-29 at 9 40 14 AM" src="https://user-images.githubusercontent.com/1269969/160663127-b5af484e-502e-43b7-83a4-6bedb98c3d53.png">

jakemac53 replied:
Oh weird... ya its orange on a chromebook :(. I am on my mac now and I see it as a greyish blue. These were both looking at the github UI, so the color is still different based on the platform and not the terminal.

That is pretty unfortunate. @munificent was suggesting using an older check mark symbol that is just the check mark (no background) and then coloring the check mark explicitly?

jakemac53 replied:
I also think we can definitely launch this with any of the suggested choices, and revist later.

devoncarew replied:
> that is just the check mark (no background) and then coloring the check mark explicitly?

That would be great if we could do it :) From experimentation, we can't use ansi colors in the summary line of a group here.

> I also think we can definitely launch this with any of the suggested choices, and revist later.

Let's do that then; I don't think there's an idea set of icons unfortunately - I've experimented w/ many. I'll change to the green checkmark, and we can iterate as needed.

jakemac53 commented on pkgs/test_core/lib/src/util/io.dart:
```diff
@@ -87,6 +87,12 @@ final _tempDir = Platform.environment.containsKey('_UNITTEST_TEMP_DIR')
+///
+/// See
+/// https://docs.github.com/en/actions/learn-github-actions/environment-variables.
+bool get githubContext => Platform.environment['GITHUB_ACTIONS'] == 'true';
```
I could call this maybe `inGithubContext` or something, it doesn't sound like a boolean to me with the current name

devoncarew replied:
👍 will do
